pkg/asset/quota/openstack/openstack.go

```diff
@@ -0,0 +1,119 @@
+	"github.com/openshift/installer/pkg/asset/installconfig/openstack/validation"
+	"github.com/openshift/installer/pkg/quota"
+	"github.com/openshift/installer/pkg/types"
+	openstackprovider "sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1"
+)
+
+// Constraints returns a list of quota constraints based on the InstallConfig.
+// These constraints can be used to check if there is enough quota for creating a cluster
+// for the install config.
+func Constraints(config *types.InstallConfig, controlPlanes []machineapi.Machine, computes []machineapi.MachineSet) []quota.Constraint {
+	ctrplConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(controlPlanes))
+	for i, m := range controlPlanes {
+		ctrplConfigs[i] = m.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	}
+	computeReplicas := make([]int64, len(computes))
+	computeConfigs := make([]*openstackprovider.OpenstackProviderSpec, len(computes))
+	for i, w := range computes {
+		computeReplicas[i] = int64(*w.Spec.Replicas)
+		computeConfigs[i] = w.Spec.Template.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec)
+	}
+
+	var ret []quota.Constraint
+	for _, gen := range []constraintGenerator{
+		controlPlane(config, ctrplConfigs),
+		compute(config, computeReplicas, computeConfigs),
+		others,
+	} {
+		ret = append(ret, gen()...)
+	}
+	return aggregate(ret)
+}
+
+func controlPlane(config *types.InstallConfig, machines []*openstackprovider.OpenstackProviderSpec) func() []quota.Constraint {
+	return func() []quota.Constraint {
+		var ret []quota.Constraint
+		for _, m := range machines {
+			ret = append(ret, machineFlavorCoresToQuota(m.Flavor, config), machineFlavorRAMToQuota(m.Flavor, config))
+		}
+		ret = append(ret, instanceConstraint(int64(len(machines))))
+		return ret
+	}
+}
+
+func compute(config *types.InstallConfig, replicas []int64, machines []*openstackprovider.OpenstackProviderSpec) func() []quota.Constraint {
+	return func() []quota.Constraint {
+		var ret []quota.Constraint
+		for idx, m := range machines {
+			coresConstraint := machineFlavorCoresToQuota(m.Flavor, config)
+			coresConstraint.Count = coresConstraint.Count * replicas[idx]
+			ramConstraint := machineFlavorRAMToQuota(m.Flavor, config)
+			ramConstraint.Count = ramConstraint.Count * replicas[idx]
+			ret = append(ret, instanceConstraint(int64(replicas[idx])), coresConstraint, ramConstraint)
+		}
+		return ret
+	}
+}
+
+func others() []quota.Constraint {
+	return []quota.Constraint{}
+}
+
+func aggregate(quotas []quota.Constraint) []quota.Constraint {
+	sort.SliceStable(quotas, func(i, j int) bool {
+		return quotas[i].Name < quotas[j].Name
+	})
+
+	i := 0
+	for j := 1; j < len(quotas); j++ {
+		if quotas[i].Name == quotas[j].Name {
+			quotas[i].Count += quotas[j].Count
+		} else {
+			i++
+			if i != j {
+				quotas[i] = quotas[j]
+			}
+		}
+	}
+	return quotas[:i+1]
+}
```

Comment:
Using a temporary map would make this easier to follow.
```suggestion
func aggregate(quotas []quota.Constraint) []quota.Constraint {
	counts := map[string]int64{}
	for _, q := range quotas {
		counts[q.Name] = counts[q.Name] + q.Count
	}
	aggregatedQuotas := make([]quota.Constraint, 0, len(counts))
	for n, c := range counts {
		aggregatedQuotas = append(aggregatedQuotas, quota.Constraint{Name: n, Count: c})
	}
	return aggregatedQuotas
}
```

Reply:
Noted; thanks for the suggestion, I'll change the patch.

Note that I re-used a function from AWS quotas:

https://github.com/openshift/installer/blob/47b4c1a046ee43bc500ff19c98a231a754cc2a70/pkg/asset/quota/aws/aws.go#L41-L58